### PR TITLE
ENH: Implement dict-like support for rename and set_names in MultiIndex

### LINKED ISSUE #20421: API: Allow MultiIndex.rename() to accept a dict as an argument
#### Code Sample, a copy-pastable example if possible

```python
In [5]: mi=pd.MultiIndex.from_product([list('abc'), list('xy'), [1,2]],names=['abc','xy','num'])

In [6]: mi.names
Out[6]: FrozenList(['abc', 'xy', 'num'])

In [7]: mi.rename(['def','v'], level=[0,2]).names
Out[7]: FrozenList(['def', 'xy', 'v'])


In [8]: mi.rename({'xy' : 'foo'})
---------------------------------------------------------------------------
# Traceback omitted

ValueError: Length of names must match number of levels in MultiIndex.
```
#### Problem description

In coordination with how `DataFrame.rename()` supports using a dictionary to rename columns, I think it would be worthwhile to allow `MultiIndex.rename()` to also support a dictionary argument.  The value of this is that you then don't need to know the level numbers of each of the index names.

If implemented, this would also help with issue #17334, since we would then implement `MultiIndex.rename` with its own docstring.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 01882ba5b4c21b0caf2e6b9279fb01967aa5d650
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.0.dev0+657.g01882ba5b
pytest: 3.4.0
pip: 9.0.1
setuptools: 38.5.1
Cython: 0.25.1
numpy: 1.14.1
scipy: 1.0.0
pyarrow: 0.8.0
xarray: None
IPython: 6.2.1
sphinx: 1.7.1
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2018.3
blosc: 1.5.1
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.0
openpyxl: 2.5.0
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.5
pymysql: 0.8.0
psycopg2: None
jinja2: 2.10
s3fs: 0.1.3
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #20421
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I tried my hand in adding support for dict-like renamings for MultiIndexes. This seemed as the most straight forward way to me, because we do not have to add complex logic as long as we can cast the dict-like objects to regular configurations.

We probably must add a docstring to MultiIndex.rename now?

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -1319,11 +1320,11 @@ def set_names(self, names, level=None, inplace: bool = False):
 
         Parameters
         ----------
-        names : label or list of label
+        names : label or list of label or dict-like for MultiIndex
             Name(s) to set.
         level : int, label or list of int or label, optional
-            If the index is a MultiIndex, level(s) to set (None for all
-            levels). Otherwise level must be None.
+            If the index is a MultiIndex and names is not dict-like, level(s) to set
```
Comment (jreback) [inline pandas/core/indexes/base.py:1326]: pls add version changed to both of these

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -1366,6 +1367,12 @@ def set_names(self, names, level=None, inplace: bool = False):
                     ( 'cobra', 2018),
                     ( 'cobra', 2019)],
                    names=['species', 'year'])
+        >>> idx.set_names({'kind': 'snake'})
```
Comment (jreback) [inline pandas/core/indexes/base.py:1370]: add a comment on this example (skip a line)

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -1376,6 +1383,12 @@ def set_names(self, names, level=None, inplace: bool = False):
         if not is_list_like(names) and level is None and self.nlevels > 1:
             raise TypeError("Must pass list-like as `names`.")
 
+        if is_dict_like(names) and not isinstance(self, ABCMultiIndex):
+            raise TypeError("Can only pass dict-like as `names` for MultiIndex.")
```
Comment (jreback) [inline pandas/core/indexes/base.py:1387]: testing these conditions?

File: pandas/core/indexes/base.py [hunk 4]
```diff
@@ -1385,6 +1398,12 @@ def set_names(self, names, level=None, inplace: bool = False):
             idx = self
         else:
             idx = self._shallow_copy()
+
+        if isinstance(self, ABCMultiIndex) and is_dict_like(names):
```
Comment (jreback) [inline pandas/core/indexes/base.py:1402]: add a comment on what this condition is doing?

File: pandas/core/indexes/base.py [hunk 5]
```diff
@@ -1385,6 +1398,12 @@ def set_names(self, names, level=None, inplace: bool = False):
...
+            level = Index(self.names).get_indexer_for(names)
```
Comment (jreback) [inline pandas/core/indexes/base.py:1403]: can you not just iterate over these to grab the levels? (and use get_level_values)?

**Review by jreback — CHANGES_REQUESTED** [review]
see comments

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1387]: Yep, in ``test_name_with_dict_like_raising``

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1370]: I've added 1.2, if this is not ready I will change to 1.3

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1326]: I've added 1.2, if this is not ready I will change to 1.3

File: pandas/core/indexes/base.py (hunk 4 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1402]: Done

File: pandas/core/indexes/base.py (hunk 5 above)
Comment (phofl) [reply pandas/core/indexes/base.py]: Yep, thx. Thats way better.

List comprehensions would be an option too, but would have to loop twice. Would prefer ist this way

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -1376,6 +1393,12 @@ def set_names(self, names, level=None, inplace: bool = False):
         if not is_list_like(names) and level is None and self.nlevels > 1:
             raise TypeError("Must pass list-like as `names`.")
 
+        if is_dict_like(names) and not isinstance(self, ABCMultiIndex):
```
Comment (jreback) [inline pandas/core/indexes/base.py:1396]: all of these conditions can be if/elif

File: pandas/core/indexes/base.py [hunk 7]
```diff
@@ -1376,6 +1393,12 @@ def set_names(self, names, level=None, inplace: bool = False):
...
+            raise TypeError("Can only pass dict-like as `names` for MultiIndex.")
+
+        if is_dict_like(names) and level is not None:
+            raise TypeError("Can not pass level when passing dict-like as `names`.")
+
```
Comment (jreback) [inline pandas/core/indexes/base.py:1401]: doesn't this hit a dict-like names?

File: pandas/core/indexes/base.py [hunk 8]
```diff
@@ -1385,6 +1408,16 @@ def set_names(self, names, level=None, inplace: bool = False):
             idx = self
         else:
             idx = self._shallow_copy()
+
+        if isinstance(self, ABCMultiIndex) and is_dict_like(names):
+            # Transform dict to list of new names and corresponding levels
+            level, names_adjusted = [], []
+            for i, name in enumerate(self.names):
```
Comment (jreback) [inline pandas/core/indexes/base.py:1415]: see my comment above i think this condition needs to be on L1402

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 6 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1396]: Wanted to be consistent with the existing conditions, changed them to elif

File: pandas/core/indexes/base.py (hunk 7 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1401]: Yikes, yes you are right. have not though about dict like names. Removed this

File: pandas/core/indexes/base.py (hunk 8 above)
Comment (phofl) [reply pandas/core/indexes/base.py:1415]: Moved it up a bit and added level is None

**Comment by jreback** [comment]
cc @Dr-Irv if you'd have a look. this is api you were expecting (from the OP)?

File: pandas/tests/indexes/multi/test_names.py [hunk 9]
```diff
@@ -148,3 +148,31 @@ def test_setting_names_from_levels_raises():
...
+        ({"x": "z", "y": "x"}, ["z", "x", "z"]),
+        ({"y": "z"}, ["x", "z", "x"]),
+        ({}, ["x", "y", "x"]),
+        ({"z": "a"}, ["x", "y", "x"]),
+        ({"y": "z", "a": "b"}, ["x", "z", "x"]),
+    ],
+)
+def test_name_mi_with_dict_like(func, rename_dict, exp_names):
+    # GH#20421
+    mi = MultiIndex.from_arrays([[1, 2], [3, 4], [5, 6]], names=["x", "y", "x"])
```
Comment (Dr-Irv) [inline pandas/tests/indexes/multi/test_names.py:167]: Can you also include a test for the use case 
```
mi = MultiIndex.from_arrays([[1, 2], [3, 4], [5, 6]], names=["x", "y", "z"])
result = mi.rename({"x":"a"})
```
In other words, start with a `MultiIndex` that has different names for the level, and make sure they rename correctly.

We also should decide if we allow duplicate level names (which your test is producing).  But maybe that's a separate issue.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Needs tests for the basic use case in the original issue.

**Comment by Dr-Irv** [comment]
> cc @Dr-Irv if you'd have a look. this is api you were expecting (from the OP)?

Yes, that is the API.  I didn't check the implementation, but the tests should reflect the "normal" use case of having a `MultiIndex` with names that are different for each level, and you want to change one or two of the names.  Also, should check that behavior is consistent with `DataFrame.rename()` if a name in the `dict` argument is not one of the level names - I'm not sure what we do in `DataFrame.rename()` if that is the case.

File: pandas/tests/indexes/multi/test_names.py (hunk 9 above)
Comment (phofl) [reply pandas/tests/indexes/multi/test_names.py:167]: Added the tests.

**Comment by phofl** [comment]
@Dr-Irv 

Concerning missing names: I have oriented me on df.rename, as you also suggested. df.rename does nothing for missing index elements or column names.

Additionally for duplicates: Same holds true here. Is also done like df.rename. It produces duplicates if we have the same name twice in the original and it also produces duplicates if the targets already exist.

**Review by Dr-Irv — APPROVED** [review]
LGTM.  Final approval rests with @jreback

**Comment by phofl** [comment]
ping

File: pandas/tests/indexes/multi/test_names.py [hunk 10]
```diff
@@ -150,3 +150,49 @@ def test_setting_names_from_levels_raises():
...
+    result = getattr(mi, func)(rename_dict)
+    expected = MultiIndex.from_arrays([[1, 2], [3, 4]], names=exp_names)
+    tm.assert_index_equal(result, expected)
+
+
+def test_name_with_dict_like_raising():
+    # GH#20421
+    ix = pd.Index([1, 2])
+    msg = "Can only pass dict-like as `names` for MultiIndex."
+    with pytest.raises(TypeError, match=msg):
```
Comment (jreback) [inline pandas/tests/indexes/multi/test_names.py:197]: can you add a test when you do pass a level= kwarg. (which should raise)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py [hunk 11]
```diff
@@ -1378,11 +1379,18 @@ def set_names(self, names, level=None, inplace: bool = False):
...
+
+            .. versionchanged:: 1.2.0
+
         level : int, label or list of int or label, optional
-            If the index is a MultiIndex, level(s) to set (None for all
-            levels). Otherwise level must be None.
+            If the index is a MultiIndex and names is not dict-like, level(s) to set
+            (None for all levels). Otherwise level must be None.
+
+            .. versionchanged:: 1.2.0
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/indexes/base.py]: Done

File: pandas/core/indexes/base.py [hunk 12]
```diff
@@ -1425,16 +1433,37 @@ def set_names(self, names, level=None, inplace: bool = False):
                     ( 'cobra', 2018),
                     ( 'cobra', 2019)],
                    names=['species', 'year'])
+
+        When renaming levels through a dictionary no level can't be passed.
```
[the remark this answers is not in the document]
Comment (phofl) [reply pandas/core/indexes/base.py]: Done

File: pandas/tests/indexes/multi/test_names.py (hunk 10 above)
Comment (phofl) [reply pandas/tests/indexes/multi/test_names.py:197]: Raises now

File: pandas/core/indexes/base.py [hunk 13]
```diff
@@ -1378,11 +1379,18 @@ def set_names(self, names, level=None, inplace: bool = False):
 
         Parameters
         ----------
-        names : label or list of label
+
+        names : label or list of label or dict-like for MultiIndex
             Name(s) to set.
+
+            .. versionchanged:: 1.2.0
```
Comment (jreback) [inline pandas/core/indexes/base.py:1386]: change to 1.3

**Review by jreback — CHANGES_REQUESTED** [review]
comment, ping on green

File: pandas/core/indexes/base.py (hunk 13 above)
Comment (phofl) [reply pandas/core/indexes/base.py]: Thx

**Comment by phofl** [comment]
@jreback green

File: pandas/core/indexes/base.py (hunk 13 above)
Comment (phofl) [reply pandas/core/indexes/base.py]: Done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**